tkg/client/update_credentials.go

```diff
@@ -197,3 +212,118 @@ func (c *TkgClient) updateVSphereCredentialsForCluster(clusterClient clusterclie
+				continue
+			}
+
+			log.V(4).Infof("Updating credentials for workload cluster %q ...", clusters[i].Name)
+			err := c.updateAzureCredentialsForCluster(clusterClient, &UpdateCredentialsOptions{
+				ClusterName:               clusters[i].Name,
+				Namespace:                 clusters[i].Namespace,
+				IsRegionalCluster:         false,
+				IsCascading:               options.IsCascading,
+				AzureUpdateClusterOptions: options.AzureUpdateClusterOptions,
+			}, false)
+			if err != nil {
+				log.Error(err, "unable to update credentials for workload cluster")
+				continue
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *TkgClient) updateAzureCredentialsForCluster(clusterClient clusterclient.Client, options *UpdateCredentialsOptions, force bool) error {
+	updateIdentity := force || options.IsCascading
+	if options.IsRegionalCluster {
+		// update capz-manager-bootstrap-credentials
+		log.Infof("Updating secret capz-manager-bootstrap-credentials for management cluster %q", options.ClusterName)
+		if err := clusterClient.UpdateCapzManagerBootstrapCredentialsSecret(options.AzureUpdateClusterOptions.AzureTenantID, options.AzureUpdateClusterOptions.AzureClientID, options.AzureUpdateClusterOptions.AzureClientSecret); err != nil {
+			return err
+		}
+
+		// restart capz-controller-manager pod
+		log.Infof("Restart capz-controller-manager pod")
+		if err := c.restartCAPZControllerManagerPod(clusterClient); err != nil {
+			return err
+		}
+	}
+
+	if updateIdentity {
+		// UpdateAzureClusterIdentity
+		log.Infof("Update cluster %q AzureClusterIdentity", options.ClusterName)
+		if err := clusterClient.UpdateAzureClusterIdentity(options.ClusterName, options.Namespace, options.AzureUpdateClusterOptions.AzureTenantID, options.AzureUpdateClusterOptions.AzureClientID, options.AzureUpdateClusterOptions.AzureClientSecret); err != nil {
+			return err
+		}
+	} else {
+		unifiedIdentity, err := clusterClient.CheckUnifiedAzureClusterIdentity(options.ClusterName, options.Namespace)
+		if err != nil {
+			return err
+		}
+		if unifiedIdentity {
+			log.Warningf("WARNING: workload cluster %s use the same AzureClusterIdentity with Management Cluster and it is only updated together with Management Cluster", options.ClusterName)
+			updateIdentity = true
+		}
+	}
+
+	if updateIdentity {
+		// Recycle all KCP in the cluster
+		log.V(4).Infof("Update KCP rolloutAfter for cluster %q", options.ClusterName)
+		if err := clusterClient.UpdateAzureKCP(options.ClusterName, options.Namespace); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *TkgClient) restartCAPZControllerManagerPod(clusterClient clusterclient.Client) error {
+	replicas, err := clusterClient.GetCAPZControllerManagerDeploymentsReplicas()
+	if err != nil {
+		return err
+	}
+
+	if replicas != 0 {
+		log.Infof("Set capz-controller-manager deployment replicas to 0")
+		if err := clusterClient.UpdateCAPZControllerManagerDeploymentReplicas(int32(0)); err != nil {
+			return err
+		}
+
```

Comment:
Is it necessary to add the check to see
* if the pods are gone when deployment replicas set to 0
* if the pods are back and deployment back to running normally
to see if the restart is successful and perhaps the updated credentials are valid?

Reply:
Yes. We should check deployment replicas using VerifyAutoscalerDeploymentAvailable